oeps/oep-0011.rst

```diff
@@ -0,0 +1,412 @@
+
+
+Open Questions
+==============
+
+.. _JavaScript frameworks:
+
+1. **Which more modern JavaScript framework should be supported?**
+
+   The JavaScript space is very fast moving as can be seen in
+   `The State of JavaScript 2016 Survey`_. There is a lot of momentum
+   around `React`_, and it appears to have many qualities that would
+   make it a good fit for edX. It is too early to make this call, but
+   plans are being made to evaluate its usage. See `React at edX`_ for
+   more details.
+
+2. **How should offline support be provided?**
+
+   For many mobile users, it would be very useful if they could continue
+   their Open edX courses while offline. This is true both for true mobile
+   apps as well as for responsive web applications. This is an area that
+   has not been explored much up to now, but which will only become more
+   important going forward. In particular, we should look at the great
+   capabilities provided by service workers. See the article
+   `Service Workers: An Introduction`_ for a good starting point.
+
+3. **Should Universal JavaScript be considered?**
+
+   There are many companies that are seeing performance benefits by
+   pre-rendering their dynamic client-side applications on the server first.
+   The most common approach is to run the same JavaScript code both in the
+   browser and using Node on the server. This requires the use of a JavaScript
+   framework that does not perform direct manipulation of the DOM, but instead
+   works with an abstraction of the desired HTML. For example, Netflix is
+   seeing great performance gains by using `React`_ in this way, as can be
+   seen in the blog post `Making Netflix Faster`_.
+
+4. **Where do web components fit with Open edX applications?**
+
+   Web Components is a W3C recommendation that allows developers to create
+   custom HTML elements for use within their web applications. The
+   `Polymer Project` provides a library that allows web components to be
+   built in most modern browsers. It seems that there is a useful overlap
+   between web components and Open edX's `XBlocks`_
+
+5. **Should HTTP/2 be used in preference to asset bundling?**
+
+   The Hypertext Transfer Protocol Version 2 (HTTP/2) enables more efficient
+   network communication between clients and servers. In particular, it
+   allows for multiple requests to be handled with the same TCP connection,
+   which removes most of the benefits of asset bundling. It is not yet
+   clear whether this is ready for full scale adoption, but it would reduce
+   a lot of complexity from application's asset pipelines.
+
+6. **Should Jinja2 be considered in preference to Django Templates?**
+
+   `Jinja2`_ is a popular alternative to Django Templates, that has not been
+   evaluated for use within Open edX.
+
+
+Rejected Alternatives
+=====================
+
+Note: this section needs fleshing out...
+
+.. _Reasons for rejecting Angular:
+1. **Angular**
+
+   Needs fleshing out...
```

Comment:
Some bullet points that might provide guidance for fleshing this out:

- it's a very heavy, opinionated framework that doesn't interop with other systems nicely. In comparison, React is known for being able to integrate with other frameworks (especially Backbone) making a migration less painful. [Here's a good talk from folks at Hubspot about their transition from Backbone to React](http://product.hubspot.com/blog/moving-backbone-to-flux-react) (and I think Airbnb went through this as well).
- Angular's community management hasn't been great. Their abandonment of Angular 1 left a lot of Angular shops in the lurch and there's a general consensus that the plans for 2 were not well thought-out (originally not planned to have any kind of upgrade path from v1 to v2, originally planned to require a totally new language called AtScript, later changed after community outrage to be a soft requirement on Typescript). There's still quite a bit of ill will and I don't trust Google to not pull the same thing again. In comparison, Facebook uses React virtually everywhere in their own apps and they have a very strong commitment to supporting it, breaking the APIs as little as possible and releasing codemods for it when they do because it hurts them internally as much as it hurts the community.
- none of our Fed developers know Angular 2 or Typescript AFAIK, by contrast we have a decent amount of React experience on staff already
- this maybe leans too heavily on my personal Opinions but I was an Angular 1 developer for a year or so and I hated it - too much reliance on framework magic, very difficult to test and debug things, terrible performance characteristics on complex pages, etc. I haven't used Angular 2 but it just seems to me that it would be hard for it to not bring that legacy along in the new version unless they fired all the people that designed the first one.